Modify_mf

fachra opened the pull request:
This pull request focuses on improving the performance of the Matrix Formalism module.

### Main changes: 
**1. If all permeabilities are zero, the Matrix Formalism module separately computes the eigenfunctions for each compartment.**
- If all permeabilities are zero, `lap_eig` becomes a indexed structure. `lap_eig(icmpt)` contains the eigvalues, eigfuncs, totaltime and length_scales for compartment icmpt. 
`compute_laplace_eig` computes the eigendecomposition for each compartment separately if all permeabilities are zero.
```diff
- lap_eig = compute_laplace_eig(femesh, setup.pde, eiglim, setup.mf.neig_max);
+ lap_eig = compute_laplace_eig(femesh, setup.pde, setup.mf, savepath);
```
- `mf_jn` becomes a cell of size `1 x ncompartment`. `mf_jn{icmpt}` contains the `jn` for compartment icmpt.
```diff
- mf_jn = compute_mf_jn(lap_eig.values, setup);
+ mf_jn = compute_mf_jn(lap_eig, setup);
```
- `diffusion_tensor` becomes an array of size [3 x 3 x nsequence x ncompartment]. `diffusion_tensor_all` is the diffusion tensor for the whole geometry. `a` is a `1 x ncompartment` cell containing the first-order moment.
```diff
- diffusion_tensor = compute_mf_diffusion_tensor(femesh, lap_eig, mf_jn);
+ [diffusion_tensor, diffusion_tensor_all, mf_jn, a] = compute_mf_diffusion_tensor(femesh, setup, lap_eig);
```

**2. Improve the computational efficiency of the Matrix Formalism module.**
- Replace `expm` with `expmv` to avoid the time-and-memory-consuming matrix exponential operation.
- Change the default `sigma` parameter of `eigs`. The new setting can bring a ~5x speed-up without degrading the precision.
```diff
- [funcs, values] = eigs(K+Q, M, neig_max, "smallestreal", otherparams{:});
+ [funcs, values] = eigs(K+Q, M, neig_max, -1e-8, otherparams{:});
```
- Remove `parfor` because the time-consuming operations, such as `eigs`, `eig`, `expm`, `expmv`, and matrix multiplication, are multithreading and can use all available cores. `parfor` only adds unnecessary overhead.
-  TODO: Support multiple GPU.

**3. Implement the surface relaxation approximation.** https://github.com/jingrebeccali/SpinDoctor/pull/3/commits/261b247424835c6f5d6af004fff3e07c0c769eb0
- Grebenkov treats the transport phenomenon across the cell membranes as surface relaxation in his paper *Laplacian Eigenfunctions in NMR. I. A Numerical Tool*. Based on this idea, there is an alternative way to implement permeable interfaces (see the section *Alternative Implementation of Uniform Surface Relaxation* in Grebenkov's paper). The advantage of the surface relaxation approximation is that it only requires the eigendecomposition with the zero Neumann boundary condition to conduct simulation with permeable interfaces. 
- The surface relaxation approximation degrades the precision. But its precision is still acceptable (relative error <= 1% using BTPDE as the reference) with appropriate truncation of the eigendecomposition.
- I add a new option `surf_relaxation` to the `setup.mf`.
```matlab
setup.mf.surf_relaxation = true;
```

**4. Optimize the I/O module for saving and loading simulated results.**
- Use folder name and md5 hash code to encode different setup settings. https://github.com/jingrebeccali/SpinDoctor/pull/3/commits/788d1eb2af17cb164e8afb683f7c08a943a7db21
- Ensure that we save the results of different settings in different folders.
- When we want to compute the Laplace eigenmodes and a larger decomposition exists (e.g. with more eigenmodes or smaller length scale), the Matrix Formalism module computes the required eigendecomposition by truncating the larger decomposition to avoid repeating the time-consuming eigendecomposition.
- Optimize the loading procedure. https://github.com/jingrebeccali/SpinDoctor/pull/3/commits/fb821a5a352f0e517bfdbe1e7cb3485da89d9ac2
- Add some loading functions to get a better loading performance. ([load_btpde](https://github.com/jingrebeccali/SpinDoctor/blob/modify_mf/src/load_btpde.m), [load_btpde_midpoint](https://github.com/jingrebeccali/SpinDoctor/blob/modify_mf/src/load_btpde_midpoint.m), [load_hadc](https://github.com/jingrebeccali/SpinDoctor/blob/modify_mf/src/load_hadc.m), [load_laplace_eig](https://github.com/jingrebeccali/SpinDoctor/blob/modify_mf/src/load_laplace_eig.m), [load_mf](https://github.com/jingrebeccali/SpinDoctor/blob/modify_mf/src/load_mf.m), [load_simulation](https://github.com/jingrebeccali/SpinDoctor/blob/modify_mf/src/load_simulation.m))

**5. Add support for PLY and OFF type meshes.** https://github.com/jingrebeccali/SpinDoctor/pull/3/commits/3d15cc2eb3d3b1309183aab8c987c7e289a80634

### Minor changes:
- add `tpause` effect of DoublePGSE for MF simulation. https://github.com/jingrebeccali/SpinDoctor/pull/3/commits/0ba7a4731b328c7d89cc6404774184766e72eb1e
- optimize MF simulation for OGSE sequences. https://github.com/jingrebeccali/SpinDoctor/pull/3/commits/759c9ccf1019dedb33b60d97d0da830222f031cd
- add `docs/` folder.
- update `CITATION.bib`.

Commit:
Minor changes

Commit:
Modify MF for uncorrelated compartments

agdestein commented on src/solve_mf.m:
```diff
@@ -63,92 +58,188 @@
+            edL = expm(-(seq.Delta - seq.delta) * (L + T2));
+
+            % Laplace coefficients of final magnetization
+            nu = edK' * (edL * (edK * (edK' * (edL * (edK * nu0)))));
```
We should probably also account for `tpause`, as diffusion also occurs between the pulses
`etL = expm(-seq.tpause * (L + T2));`
`nu = edK' * (edL * (edK * (etL * (edK' * (edL * (edK * nu0))))));`

agdestein replied:
Also, for these exponential time stepping schemes, we may consider not explicitly computing the matrix exponentials at all. A MATLAB function `expmv` is provided here (BSD-2 license), [https://github.com/higham/expmv](https://github.com/higham/expmv), so that instead of writing `exp(-delta * K) * vector` we can write `expmv(-delta, K, vector)`

fachra replied:
Thanks, Syver! I think you are right, we should account for `tpause`. We should also change Eq.36 in the UserGuide as well.

I actually tried `expmv` in another project. It seems that the accuracy of `expmv` is not very satisfactory, especially with a large matrix. But probably I used `expmv` in the wrong way. I will work on this point later. Because it could greatly improve computational efficiency.

agdestein replied:
Yes, it is an approximation, but that is also the case for the standard `expm` I believ (which uses the scaling and squaring method). It is probably a good idea to check if the accuracy is satisfactory and under what conditions before using such an approach

fachra replied:
Good idea! I will check them.

fachra replied:
In all cases I have tested so far, `expmv` can give a satisfactory precision. I have replaced `expm` with `expmv`. For large matrices, `expmv` is much faster than `expm`. However, `expm` seems faster for relatively small matrices.

`expmv` is especially important for MF simulation of CosOGSE and SinOGSE which requires 100+ `expmv` operations.

Commit:
fix bug: add tapuse effect of DoublePGSE for MF simulation

ps: made minor changes for the string of Sequence

Commit:
replace expm with expmv

Commit:
plot eigenfunctions for a certain compartment

other changes: modified the way to save and load eigendecomposition

fachra committed:
set massrelax to 0 when all relaxations are Inf

fachra committed:
reset lap_eig based on a new eiglim

Commit:
update UserGuide and add folder docs

Commit:
optimize the computation of lap coefficients

fachra commented on src/solve_mf.m:
```diff
@@ -63,92 +58,217 @@
+
+            % Laplace coefficients of final magnetization
+            nu = expmv(-seq.delta, K, nu0);
+            nu = expm(-(seq.Delta - seq.delta) * (L + T2)) * nu;
```
In general, `expmv` is faster than `expm`. But for some unknown reasons, `expmv(-t, L+T2, nu)` can only use one process and is much slower than `expm(-t*(L+T2)*nu` which seems capable of leveraging the structure of `L+T2`.
I tested both zero and non-zero `T2` matrices, `expmv(-t, L+T2, nu)` is always very slow.

fachra commented on src/solve_mf.m:
```diff
@@ -63,92 +58,217 @@
+            edK = expm(-seq.delta * K);
+            edL = expm(-(seq.Delta - seq.delta) * (L + T2));
+            etL = expm(-seq.tpause * (L + T2));
+            nu = edK' * (edL * (edK * (etL * (edK' * (edL * (edK * nu0))))));
```
Since `edK` is reused several times and the computation of `edL` and `etL` is quite fast, the way of using `expm` is actually faster. The drawback is that `expm` needs more memory.

Commit:
refactor solve_mf

Commit:
refactor IO pipeline, drivers and solvers

no change for karger and analytical methods

Commit:
fix bugs

Commit:
replace md5 hash function

Commit:
add new driver, setups and test

Commit:
fix bug

Commit:
fix inconsistent md5 hash code

fix the issue that the md5 hash code for the same simulation are different due to different machine or matlab version

Commit:
Optimize loading performance

Commit:
support char type setup.name

Operations like "+" do not support char type setup.name

Commit:
simplify the computation when b=0

when b=0, the signal is the same in all direction

Commit:
rewrite save/load_simulation output

Commit:
optimize MF simulation for OGSE sequences

Commit:
reduce gradient-direction saving precision

To avoid inconsistency between different machines

Commit:
minor change

Commit:
Merge branch 'modify_mf' of github.com:jingrebeccali/SpinDoctor into modify_mf

Commit:
add support for PLY and OFF type meshes

Commit:
add error messages to the preparation step

Commit:
change the save path

Commit:
improve saving and loading performance

Commit:
update citation

Commit:
fix minor bugs

Commit:
fix string format

Commit:
fix a bug about saving mf results

Commit:
revise loading procedure in solve_mf

Commit:
add surface relaxation approximation

The surface relaxation approximation implements and extends the idea of the alternative implementation of uniform surface relaxation proposed by Grebenkov in the paper Laplacian Eigenfunctions in NMR I A Numerical Tool.

Commit:
grant tetgen executable permission in macOS

Commit:
shorten folder and file names

windows system has a very strict constraint for the total length of a path (< 260 characters)

Commit:
add precompiled hash function

Commit:
fix bug in load_mf

Commit:
add warning about mf and fix bugs

matrix formalism method is inaccurate for nonequilibrium spin density distribution

Commit:
reduce tight_wrap ECS vertices number

Commit:
set default parfor MaxNumWorkers to 2048 instead of 12

Commit:
check if Parallel Computing Toolbox is licensed

Commit:
fix the warning about eigenvalues

jingrebeccali committed:
Prepare change branch name

jingrebeccali committed:
Prepare change branch name
